### ISSUE TITLE
Arrow keys don't work in vim while using bash from Cmder.

### ISSUE THREAD

**Issue by dsouzadyn** [Microsoft/BashOnWindows#1154]
I discovered an issue when using Vim. It does not allow me to use the arrow keys at all and gives me an error in vim. This issue only occurs when you use bash from within Cmder. I'm using Cmder which uses ConEmu version 160914 [64].
![capture](https://cloud.githubusercontent.com/assets/5621469/19021178/789ae0da-88d9-11e6-9f3b-882a9aa7a8e1.JPG)

Reproduce using the following steps:
1. Start Cmder.
2. Type in ```bash``` (unless it's your default shell).
3. Open a file in vim with ```vim <filename>```.
4. Type in anything and try to move around with the arrows.

**Comment by fpqc** [comment]
@dsouzadyn  The arrow keys are bound to diff escape sequences.  You need to use a special mode in cmder or conemu.

**Comment by rmobis** [comment]
See https://github.com/cmderdev/cmder/issues/901 and this `[ConEmu release](https://conemu.github.io/blog/2016/04/11/Build-160411.html)` for a solution.

**Comment by fpqc** [comment]
Yes you need -cur_console:p

**Comment by zadjii-msft** [comment]
I'm going to close this, as it's a conemu/cmder problem, and not a WSL/conhost one

**Comment by AuthorProxy** [comment]
Guys, how can I echo something or run any other command after zsh has loaded?
I was tried:
```%windir%\System32\bash.exe ~ -c zsh & echo test -new_console:pad:C:\ProxyDocuments\Projects```
but this doesn't work

Also why zsh git is so slow..

**Comment by zadjii-msft** [comment]
@AuthorProxy Could you open a new issue for this? I'd rather not pollute an old, closed thread.

**Comment by fpqc** [comment]
@zadjii-msft Also his question should go to the ConEmu issues page or go to @Maximus5 on twitter.

**Comment by Maximus5** [comment]
This is not a bug in ConEmu. This is a bug of conhost/BashOnWindows, mentioned here long ago
https://github.com/Microsoft/BashOnWindows/issues/111

ConEmu tries to provide workaround for this conhost bug, but due to the lack of API its impossible for every possible case.

**Comment by Maximus5** [comment]
It also depends on another bug: https://github.com/Microsoft/BashOnWindows/issues/406

**Comment by fpqc** [comment]
@Maximus5 Yeah, but they said they're working on a proper new console API now. So in the meantime, since you have workarounds, there is no point sending people here.

**Comment by DanielGGordon** [comment]
What is the current status of the latest Windows build working with Cmder? 
1. Colors
2. Arrow keys
3. Zsh?

I'm very excited to one day have a Windows machine with the new Bash on Windows - inside Cmder, with working keys and colors - and oh-my-zsh. That would just be so awesome.

**Comment by zadjii-msft** [comment]
I wouldn't expect any more improvements in the coming release in terms of supporting conemu/cmder. It's on our radar, but it's not going to make it for this release.

**Comment by luozixiao1996** [comment]
@DanielGGordon 
emmm Seems it works very well on Version 1709 (with some small bug fix tricks)

**Colors**
![snipaste_2017-10-28_02-46-05](https://user-images.githubusercontent.com/13526084/32120358-d33f68dc-bb8a-11e7-8a72-2e36a691addc.png)

**Arrow keys**
Temporary fix this by adding
`
    function zsh () {
        bash -cur_console:p
    }
`
to %CmderDIR%/vendor/profile.ps1 and "zsh" in Cmder (Powershell, for cmd it is the clink.lua)
* This first start bash, and I've modified the .bashrc to let it switch to zsh automatically.
Add this to $Home/.bashrc:
`
# if running in terminal...
if test -t 1; then
# ...start zsh
exec zsh
fi
`

**(oh-my-)zsh**
As the picture above : )

**Comment by gsusI** [comment]
This worked for me `echo ':set term=builtin_ansi' >> ~/vimrc`

**Comment by sarlaccpit** [comment]
> This worked for me `echo ':set term=builtin_ansi' >> ~/vimrc`

worked for me to.   Just has to be `~/.vimrc`

**Comment by ferenczy** [comment]
> echo ':set term=builtin_ansi'

Interesting. In my case, it affected syntax highlighting colors but mainly it hid the text cursor in Vim so it's not usable for me, unfortunately. You can't really edit a text if you don't see where the cursor is.

Ubuntu 18.04 on Windows 10 Enterprise version 1809

**Comment by ferenczy** [comment]
It took me some time and many unsuccessful attempts, but finally I managed to fix the issue while not breaking colors or any other things!

Instead of changing terminals and modes, I focused on key codes generated by the cursor keys and what Vim has mapped to the cursor movement.

## TL;DR

To fix the cursor movement using the arrow keys, add the following commands into the `.vimrc` file:

    set <Up>=^[[A
    set <Down>=^[[B
    set <Right>=^[[C
    set <Left>=^[[D

Tested in Vim 8.0 under Ubuntu 18.04 on Windows 10 Enterprise version 1809.

## Explanation

My goal was to find out what actual key codes the arrow keys are generating on my system and what key codes has Vim mapped to its cursor movement actions.

### Key codes assigned to cursor movement actions

In Vim, execute `:set <key-name>?` to print the key code currently mapped to a specific key name. You can find the key names in [this list](http://vimdoc.sourceforge.net/htmldoc/term.html#t_ku) or by switching Vim to the *Insert* mode and pressing <kbd>Ctrl</kbd>+<kbd>K</kbd> followed by the actual key. So, for example, for the left arrow key, the name is `<Left>`, so I executed the command `:set <Left>?` to get the following output:

    t_kl <Left>      ^[O*D

That means, the key named `<Left>`, which [means `cursor-left`](http://vimdoc.sourceforge.net/htmldoc/intro.html#cursor-left), is mapped to the [terminal option `t_kl`](http://vimdoc.sourceforge.net/htmldoc/term.html#t_kl) and has assigned the key code `^[O*D`.

### Key codes generated by keyboard

To get the actual key codes generated by keys, switch to the *Insert* mode, press <kbd>Ctrl</kbd>+<kbd>V</kbd> and then the key. For example, for the left arrow key, it prints the code `^[[D` on my system, which is clearly different from the key code `^[O*D` assigned to the key name `<Left>`.

### Fix

So what's needed is simply to map Vim's key names to the same key codes as generated by the keyboard:

    :set <Left>=^[[D

Then I repeated the process for rest of the arrow keys and finally added the following commands to the `.vimrc` file:

    set <Up>=^[[A
    set <Down>=^[[B
    set <Right>=^[[C
    set <Left>=^[[D

Now, the cursor keys work in all Vim modes.

**Comment by satriajanaka09** [comment]
> 
> 
> This worked for me `echo ':set term=builtin_ansi' >> ~/.vimrc`

This worked for me as well. I use ConEmu. `set nocompatible` on both `.vimrc` and `.xhrc` did not work. I only added `:set term=builtin_ansi` in `.vimrc` and now the arrow key works as I expected.
